llvm/lib/CodeGen/BasicBlockSections.cpp

```diff
@@ -265,8 +265,22 @@ void llvm::sortBasicBlocksAndUpdateBranches(
 // zero implies "no landing pad." This function inserts a NOP just before the EH
 // pad label to ensure a nonzero offset.
 void llvm::avoidZeroOffsetLandingPad(MachineFunction &MF) {
+  std::optional<MBBSectionID> CurrentSection;
+  bool SectionEmpty = true;
+  auto IsFirstNonEmptyBBInSection = [&](const MachineBasicBlock &MBB) {
```

Comment:
This function may give false positives in the following situation:
```
bb0 (bbsections 0)
bb1 (bbsections cold)
bb2 (bbsections 0)
```

If false positives are fine (since we're just inserting a `nop`), then you can disregard this feedback. Otherwise, you may want to do something like
```
llvm::DenseSet<MBBSectionID> SectionsWithNonEmptyBlocks;
auto IsFirstNonEmptyBBInSection = [&](const MachineBasicBlock &MBB) {
  if (MBB.empty() || SectionsWithNonEmptyBlocks.contains(MBB.getSectionID()))
    return false;

  SectionsWithNonEmptyBlocks.insert(MBB.getSectionID());
```

Reply:
Good catch! I was under the impression that basic blocks are assumed to be sorted. Both callsites `avoidZeroOffsetLandingPad` are preceded by a call to `sortBasicBlocksAndUpdateBranches`. However, I see now that `finishAdjustingBasicBlocksAndLandingPads` sorts only by section type, but not by section number. I am not sure how section numbers are used.

Another thing I noticed is that `sortBasicBlocksAndUpdateBranches` calls `assignBeginEndSections`. If this issue already exists where sections are interleaved with each other after sorting, this would incorrectly mark some blocks as section begin as well. But that also means `avoidZeroOffsetLandingPad` should probably keep using `MBB.isBeginSection()`, to not duplicate that behavior if it changes in future.

Reply:
Ahh, no you're right, `sortBasicBlocksAndUpdateBranches` guarantees the basic block sections will be contiguous. Looks good!